File: transports/typescript/test/conformance.test.ts
// Verify the TypeScript reference library against the checked-in conformance
// vectors — the normative, language-neutral byte-exact contract. This is a
// faithful port of `transports/rust/tests/conformance.rs`: for every vector,
// reconstruct the envelope from its `input`, assert encode → `hex`, and assert
// decode(`hex`) → an equal envelope. The vectors win; if our bytes differ, our
// code is wrong.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";

import { bytesToHex, hexToBytes, Status } from "../src/conventions.ts";
import { RpcRequest, RpcResponse, RpcPush } from "../src/rpc.ts";
import { Event, Hello, HelloAck, Heartbeat, Close, type Profile } from "../src/events.ts";
import { Datagram, CompactDatagram } from "../src/datagrams.ts";

interface Vector {
  name: string;
  description: string;
  hex: string;
  // language-neutral fields; shape varies per transport, hence `any`.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  input: Record<string, any>;
}

function load(name: string): Vector[] {
  const path = join(import.meta.dirname, "..", "..", "conformance", name);
  const doc = JSON.parse(readFileSync(path, "utf8")) as { vectors: Vector[] };
  return doc.vectors;
}

// null → undefined, so a missing optional field is the absence of a value rather
// than a present null. Mirrors how the Rust `opt_*` helpers map JSON null.
function opt<T>(v: T | null | undefined): T | undefined {
  return v === null || v === undefined ? undefined : v;
}

test("rpc vectors", () => {
  for (const vec of load("rpc.json")) {
    const { name, input } = vec;
    const payload = () => hexToBytes(input.payload_hex as string);
    let actual: Uint8Array;
    switch (input.kind) {
      case "request": {
        const r = new RpcRequest(input.service as string, input.op as string, payload());
        r.id = opt<number>(input.id);
        r.auth = opt<string>(input.auth);
        actual = r.encode();
        assert.deepStrictEqual(RpcRequest.decode(actual), r, `decode ${name}`);
        break;
      }
      case "response": {
        const r = new RpcResponse(input.status as number, payload());
        r.id = opt<number>(input.id);
        r.variant = opt<string>(input.variant);
        r.error = opt<string>(input.error);
        actual = r.encode();
        assert.deepStrictEqual(RpcResponse.decode(actual), r, `decode ${name}`);
        break;
      }
      case "push": {
        const p = new RpcPush(input.service as string, input.event as string, payload());
        actual = p.encode();
        assert.deepStrictEqual(RpcPush.decode(actual), p, `decode ${name}`);
        break;
      }
      default:
        throw new Error(`unknown rpc kind ${input.kind}`);
    }
    assert.equal(bytesToHex(actual), vec.hex, `encode ${name}`);
  }
});

test("events vectors", () => {
  for (const vec of load("events.json")) {
    const { name, input } = vec;

    // Control payloads are keyed by "control"; application events by "profile".
    if (typeof input.control === "string") {
      let bytes: Uint8Array;
      switch (input.control) {
        case "hello": {
          const h = new Hello(
            (input.versions as number[]).slice(),
            (input.profiles as string[]).slice(),
            opt<string>(input.service),
            opt<string>(input.auth),
          );
          bytes = h.encode();
          assert.deepStrictEqual(Hello.decode(bytes), h, `decode ${name}`);
          break;
        }
        case "hello_ack": {
          const h = new HelloAck(input.v as number, input.profile as string, opt<string>(input.session));
          bytes = h.encode();
          assert.deepStrictEqual(HelloAck.decode(bytes), h, `decode ${name}`);
          break;
        }
        case "ping": {
          const h = new Heartbeat(input.nonce as number, opt<number>(input.at));
          bytes = h.encode();
          assert.deepStrictEqual(Heartbeat.decode(bytes), h, `decode ${name}`);
          break;
        }
        case "close": {
          const c = new Close(input.status as number, opt<string>(input.reason));
          bytes = c.encode();
          assert.deepStrictEqual(Close.decode(bytes), c, `decode ${name}`);
          break;
        }
        default:
          throw new Error(`unknown control ${input.control}`);
      }
      assert.equal(bytesToHex(bytes), vec.hex, `encode ${name}`);
      continue;
    }

    const profile = input.profile as Profile;
    const payload = hexToBytes(input.payload_hex as string);
    let ev: Event;
    if (profile === "verbose") {
      ev = Event.verbose(opt<string>(input.service), input.event as string, payload);
      ev.id = opt<number>(input.id);
    } else {
      ev = Event.compact(input.service_ord as number, input.op_ord as number, payload);
      const id = opt<number>(input.id);
      if (id !== undefined) ev.id = id;
    }
    const bytes = ev.encode(profile);
    assert.equal(bytesToHex(bytes), vec.hex, `encode ${name}`);
    assert.deepStrictEqual(Event.decode(bytes, profile), ev, `decode ${name}`);
  }
});

test("datagram vectors", () => {
  for (const vec of load("datagrams.json")) {
    const { name, input } = vec;
    switch (input.profile) {
      case "cbor-array": {
        const d = new Datagram(
          input.op_ord as number,
          input.seq as number,
          hexToBytes(input.payload_hex as string),
        );
        const bytes = d.encode();
        assert.equal(bytesToHex(bytes), vec.hex, `encode ${name}`);
        assert.deepStrictEqual(Datagram.decode(bytes), d, `decode ${name}`);
        break;
      }
      case "compact-header": {
        const d = new CompactDatagram(
          input.op_ord as number,
          input.seq as number,
          hexToBytes(input.body_hex as string),
        );
        const epoch = opt<number>(input.epoch);
        if (epoch !== undefined) d.withEpoch(epoch);
        const bytes = d.encode();
        assert.equal(bytesToHex(bytes), vec.hex, `encode ${name}`);
        assert.deepStrictEqual(CompactDatagram.decode(bytes), d, `decode ${name}`);
        break;
      }
      default:
        throw new Error(`unknown datagram profile ${input.profile}`);
    }
  }
});

// Guard rails the registry naming as a sanity check that the codes line up.
test("status registry codes", () => {
  assert.equal(Status.Ok, 0);
  assert.equal(Status.VersionUnsupported, 5);
  assert.equal(Status.DeadlineExceeded, 8);
});

File: transports/elixir/test/conformance_test.exs
defmodule Csilgen.Transport.Vectors do
  @moduledoc false
  # Compiled before the test module so the module-level `for vec <- load(...)`
  # comprehensions can call it: a module cannot invoke its own functions while it
  # is still being compiled, so the loader has to live in a separate module.
  @conformance_dir Path.join(__DIR__, "../../conformance")

  def load(name) do
    @conformance_dir
    |> Path.join(name)
    |> File.read!()
    |> JSON.decode!()
    |> Map.fetch!("vectors")
  end
end

defmodule Csilgen.Transport.ConformanceTest do
  @moduledoc """
  Verify the library against the checked-in, byte-exact conformance vectors —
  the contract every transport library follows. For each vector: rebuild the
  envelope from its language-neutral `input`, assert encode → `hex`, and assert
  decode(`hex`) → the same envelope. Ported from
  transports/rust/tests/conformance.rs and transports/go/conformance_test.go.

  JSON is parsed with the OTP-native `JSON` module (Elixir 1.18+/OTP 27+), so no
  Jason dependency is needed; hex uses stdlib `Base`.
  """

  use ExUnit.Case, async: true

  alias Csilgen.Transport.{Datagrams, Events, RPC, Status}
  import Csilgen.Transport.Vectors, only: [load: 1]

  defp unhex(nil), do: <<>>
  defp unhex(""), do: <<>>
  defp unhex(hex), do: Base.decode16!(hex, case: :lower)

  defp tohex(bytes), do: Base.encode16(bytes, case: :lower)

  # JSON `null` (and an absent key) both mean ABSENT for an optional field.
  defp opt(input, key), do: Map.get(input, key)

  describe "rpc vectors" do
    for vec <- load("rpc.json") do
      @vec vec
      test "rpc/#{vec["name"]}" do
        input = @vec["input"]
        payload = unhex(input["payload_hex"])

        {actual, decoded, expected_env} =
          case input["kind"] do
            "request" ->
              env = %RPC.Request{
                service: input["service"],
                op: input["op"],
                id: opt(input, "id"),
                auth: opt(input, "auth"),
                payload: payload
              }

              {:ok, dec} = RPC.decode_request(RPC.encode_request(env))
              {RPC.encode_request(env), dec, env}

            "response" ->
              env = %RPC.Response{
                id: opt(input, "id"),
                status: input["status"],
                variant: opt(input, "variant"),
                error: opt(input, "error"),
                payload: payload
              }

              {:ok, dec} = RPC.decode_response(RPC.encode_response(env))
              {RPC.encode_response(env), dec, env}

            "push" ->
              env = %RPC.Push{
                service: input["service"],
                event: input["event"],
                payload: payload
              }

              {:ok, dec} = RPC.decode_push(RPC.encode_push(env))
              {RPC.encode_push(env), dec, env}
          end

        assert tohex(actual) == @vec["hex"]
        assert decoded == expected_env
      end
    end
  end

  describe "events vectors" do
    for vec <- load("events.json") do
      @vec vec
      test "events/#{vec["name"]}" do
        input = @vec["input"]

        case input["control"] do
          nil ->
            payload = unhex(input["payload_hex"])
            {:ok, profile} = Events.parse_profile(input["profile"])

            env =
              case profile do
                :verbose ->
                  %Events.Event{
                    service: opt(input, "service"),
                    event: input["event"],
                    id: opt(input, "id"),
                    payload: payload
                  }

                :compact ->
                  %Events.Event{
                    service_ord: input["service_ord"],
                    op_ord: input["op_ord"],
                    id: opt(input, "id"),
                    payload: payload
                  }
              end

            actual = Events.encode_event(env, profile)
            assert tohex(actual) == @vec["hex"]
            {:ok, decoded} = Events.decode_event(actual, profile)
            assert decoded == env

          control ->
            actual = encode_control(control, input)
            assert tohex(actual) == @vec["hex"]
        end
      end
    end
  end

  defp encode_control("hello", input) do
    Events.encode_hello(%Events.Hello{
      versions: input["versions"],
      profiles: input["profiles"],
      service: opt(input, "service"),
      auth: opt(input, "auth")
    })
  end

  defp encode_control("hello_ack", input) do
    Events.encode_hello_ack(%Events.HelloAck{
      v: input["v"],
      profile: input["profile"],
      session: opt(input, "session")
    })
  end

  defp encode_control("ping", input) do
    Events.encode_heartbeat(%Events.Heartbeat{nonce: input["nonce"], at: opt(input, "at")})
  end

  defp encode_control("close", input) do
    Events.encode_close(%Events.Close{status: input["status"], reason: opt(input, "reason")})
  end

  describe "datagram vectors" do
    for vec <- load("datagrams.json") do
      @vec vec
      test "datagrams/#{vec["name"]}" do
        input = @vec["input"]

        case input["profile"] do
          "cbor-array" ->
            env = %Datagrams.Datagram{
              op_ord: input["op_ord"],
              seq: input["seq"],
              payload: unhex(input["payload_hex"])
            }

            actual = Datagrams.encode_datagram(env)
            assert tohex(actual) == @vec["hex"]
            {:ok, decoded} = Datagrams.decode_datagram(actual)
            assert decoded == env

          "compact-header" ->
            base = %Datagrams.CompactDatagram{
              op_ord: input["op_ord"],
              seq: input["seq"],
              body: unhex(input["body_hex"])
            }

            env =
              case opt(input, "epoch") do
                nil -> base
                e -> Datagrams.with_epoch(base, e)
              end

            actual = Datagrams.encode_compact(env)
            assert tohex(actual) == @vec["hex"]
            {:ok, decoded} = Datagrams.decode_compact(actual)
            assert decoded == env
        end
      end
    end
  end

  test "status registry names round-trip the registry codes" do
    assert Status.name(Status.ok()) == "ok"
    assert Status.name(Status.unknown_service_or_op()) == "unknown-service-or-op"
    assert Status.name(99) == "other"
    assert Status.ok?(Status.ok())
    refute Status.ok?(Status.internal())
  end
end
